lib/cli_common/cli_common/mercurial.py

```diff
@@ -22,14 +38,40 @@ def robust_checkout(repo_url, repo_dir, branch=b'tip', revision=None):
                                 purge=True,
                                 sharebase=shared_dir,
                                 branch=branch)
-
-    cmd.insert(0, hglib.HGPATH)
-    proc = hglib.util.popen(cmd)
-    out, err = proc.communicate()
-    if proc.returncode:
-        raise hglib.error.CommandError(cmd, proc.returncode, out, err)
+    hg_run(cmd)
 
     hg = hglib.open(repo_dir)
     hg.update(rev=revision, clean=True)
 
     return hg
+
+
+def batch_checkout(repo_url, repo_dir, revision=b'tip', batch_size=100000):
+    '''
+    Helper to clone a mercurial repository using several steps
+    to minimize memory footprint and stay below 1Gb of RAM
+    It's used on Heroku small dynos
+    '''
+    assert isinstance(revision, bytes)
+    assert isinstance(batch_size, int)
+    assert batch_size > 1
+
+    log.info('Batch checkout', url=repo_url, dir=repo_dir)
+    cmd = hglib.util.cmdbuilder('clone',
+                                repo_url,
+                                repo_dir,
+                                noupdate=True,
+                                stream=True)
+    hg_run(cmd)
+    log.info('Inital clone finished')
+
+    repo = hglib.open(repo_dir)
+    target = int(repo.identify(rev=revision, num=True).strip().decode('utf-8'))
+    log.info('Target revision for incremental checkout', revision=target)
+
+    steps = list(range(1, target, batch_size)) + [target]
+    for rev in steps:
```

Comment:
Nit: shouldn't something like `for rev in range(1, target + 1, batch_size)` work?

Reply:
Not in every case i'm afraid

```
>>> list(range(1, 1000+1, 100))
[1, 101, 201, 301, 401, 501, 601, 701, 801, 901]
```